File: android/src/main/java/com/retu/asr/asrifly/IflySpeechRecognizer.java
package com.retu.asr.asrifly;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import androidx.annotation.Nullable;
import com.iflytek.cloud.ErrorCode;
import com.iflytek.cloud.InitListener;
import com.iflytek.cloud.RecognizerListener;
import com.iflytek.cloud.RecognizerResult;
import com.iflytek.cloud.SpeechConstant;
import com.iflytek.cloud.SpeechError;
import com.iflytek.cloud.SpeechRecognizer;
import com.iflytek.cloud.SpeechUtility;
import com.retu.asr.core.RecognitionListener;
import com.retu.asr.core.RecognitionOptions;
import com.retu.asr.core.RecognitionResult;
import com.retu.asr.util.AppUtil;
import java.util.Map;

/**
 * 讯飞语音识别器
 *
 * @author YangJiang
 * @date 2020/05/19
 */
public class IflySpeechRecognizer implements com.retu.asr.core.SpeechRecognizer,
    RecognizerListener {

    private final String TAG = "IflySpeechRecognizer";

    private final SpeechRecognizer mRecognizer;
    // Common recognizer listener adaptor
    private RecognitionListener mListener;
    private StringBuffer mCurResult = new StringBuffer();

    /**
     * 初始化讯飞识别模块
     *
     * @param context 运行上下文
     */
    public static void initUtility(Context context) {
        SpeechUtility.createUtility(context, "appid=" + AppUtil.getMetaDataValue(context, "asr.ifly.appid"));
    }

    /**
     * 语音识别器构造函数
     *
     * @param context 运行上下文
     * @param listener 语音识别监听器
     * @param options 可选参数
     */
    public IflySpeechRecognizer(Context context, RecognitionListener listener,
        @Nullable RecognitionOptions options) {
        mListener = listener;
        mRecognizer = SpeechRecognizer.createRecognizer(context, new InitListener() {
            @Override
            public void onInit(int code) {
                Log.d(TAG, "SpeechRecognizer init() code = " + code);
                if (code != ErrorCode.SUCCESS) {
                    RecognitionResult rst = new RecognitionResult();
                    rst.setType(code);
                    rst.setErrorResult("Init speech recognizer is failed!");
                    mListener.onError(rst);
                }
            }
        });
        setDefaultOptions();
        setOptions(options);
    }

    @Override
    public void start(@Nullable RecognitionOptions options) {
        setOptions(options);
        reset();
        int res = mRecognizer.startListening(this);
        if (res != ErrorCode.SUCCESS) {
            RecognitionResult rst = new RecognitionResult();
            rst.setType(res);
            rst.setErrorResult("Start recognizer is failed!");
            mListener.onError(rst);
        }
    }

    @Override
    public void cancel() {
        mRecognizer.cancel();
    }

    @Override
    public void stop() {
        mRecognizer.stopListening();
    }

    @Override
    public boolean destroy() {
        if (mRecognizer != null) {
            mRecognizer.cancel();
            return mRecognizer.destroy();
        }
        return true;
    }

    /**
     * 重置参数
     */
    private void reset() {
        cancel();
        // 清空临时识别结果文本缓存对象
        if (mCurResult.length() > 0) {
            mCurResult.delete(0, mCurResult.length());
        }
    }

    /**
     * 设置默认参数选项
     */
    private void setDefaultOptions() {
        // 清空参数
        mRecognizer.setParameter(SpeechConstant.PARAMS, null);
        // 设置听写引擎
        mRecognizer.setParameter(SpeechConstant.ENGINE_TYPE, SpeechConstant.TYPE_CLOUD);
        // 设置返回结果格式
        mRecognizer.setParameter(SpeechConstant.RESULT_TYPE, "plain");
        // 设置语言输入语言
        mRecognizer.setParameter(SpeechConstant.LANGUAGE, "zh_cn");
        // 设置语言区域，即简体中文下的方言
        mRecognizer.setParameter(SpeechConstant.ACCENT, "mandarin");

        // 设置语音前端点:静音超时时间，即用户多长时间不说话则当做超时处理
        mRecognizer.setParameter(SpeechConstant.VAD_BOS, "4000");
        // 设置语音后端点:后端点静音检测时间，即用户停止说话多长时间内即认为不再输入， 自动停止录音
        mRecognizer.setParameter(SpeechConstant.VAD_EOS, "1000");
        // 设置标点符号,设置为"0"返回结果无标点,设置为"1"返回结果有标点
        mRecognizer.setParameter(SpeechConstant.ASR_PTT, "1");
//        // 设置音频保存路径，保存音频格式支持pcm、wav，设置路径为sd卡请注意WRITE_EXTERNAL_STORAGE权限
//        mRecognizer.setParameter(SpeechConstant.AUDIO_FORMAT,"wav");
//        mRecognizer.setParameter(SpeechConstant.ASR_AUDIO_PATH, Environment.getExternalStorageDirectory()+"/msc/iat.wav");
    }

    /**
     * 设置可选参数
     *
     * @param options 需要设置的参数
     */
    private void setOptions(@Nullable RecognitionOptions options) {
        if (options == null) {
            return;
        }
        for (Map.Entry<String, Object> option : options.listOptions()) {
            Object optionValue = option.getValue();
            if (optionValue instanceof String) {
                mRecognizer.setParameter(option.getKey(), (String) optionValue);
            } else {
                Log.i(TAG, "Illegal parameter value for " + option.getKey());
            }
        }
    }

    /*--------------------RecognizerListener------------------------*/

    @Override
    public void onVolumeChanged(int volume, byte[] bytes) {
        mListener.onVolumeChange(volume);
    }

    @Override
    public void onBeginOfSpeech() {
        mListener.onRecordingStart();
    }

    @Override
    public void onEndOfSpeech() {
        mListener.onRecordingStop();
    }

    @Override
    public void onResult(RecognizerResult results, boolean isLast) {
        // json格式
//        String text = JsonParser.parseIatResult(results.getResultString());
        // plain格式
        String text = results.getResultString();
        Log.d(TAG, "onResult->" + text);
        mCurResult.append(text);
        if (isLast) {
            RecognitionResult rst = new RecognitionResult();
            rst.setFinalResult(mCurResult.toString());
            mListener.onResult(rst);
        }
    }

    @Override
    public void onError(SpeechError speechError) {
        RecognitionResult rst = new RecognitionResult();
        rst.setType(speechError.getErrorCode());
        // 返回部分内容
        if (mCurResult.length() > 0) {
            rst.setFinalResult(mCurResult.toString());
            mListener.onResult(rst);
            // 返回错误信息
        } else {
            rst.setErrorResult(speechError.getPlainDescription(true));
            mListener.onError(rst);
        }
    }

    @Override
    public void onEvent(int i, int i1, int i2, Bundle bundle) {
        // do nothing
    }
}

File: ios/asrifly/IFLYSpeechRecognizer.m
//
//  IFLYSpeechRecognizer.m
//  RNAsr
//
//  Created by YangJiang on 2020/5/18.
//

#import "IFLYSpeechRecognizer.h"
#import "IFlyMSC/IFlyMSC.h"
#import "ISRDataHelper.h"
#import "IATConfig.h"

/**
 * 记录当前的讯飞识别套件是否初始化，其只需要初始化一次
 */
static BOOL IS_INIT = FALSE;

@interface IFLYSpeechRecognizer()<IFlySpeechRecognizerDelegate>

@property (nonatomic, strong) IFlySpeechRecognizer *iFlySpeechRecognizer;
@property (nonatomic, strong) id<RecognitionListener> listener;
@property (nonatomic, strong) NSMutableString *curResult; // the results of current session

@end

@implementation IFLYSpeechRecognizer

#pragma mark - static methods
/**
 * 初始化讯飞识别套件
 */
+ (void)initUtility {
    if (IS_INIT) {
        return;
    }
    IS_INIT = TRUE;
    
    //Set log level
    [IFlySetting setLogFile:LVL_ALL];

    //Set whether to output log messages in Xcode console
    [IFlySetting showLogcat:YES];

    //Set the local storage path of SDK
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
    NSString *cachePath = [paths objectAtIndex:0];
    [IFlySetting setLogFilePath:cachePath];

    //Set APPID
    NSString *initString = [[NSString alloc] initWithFormat:@"appid=%@", [IATConfig appId]];
    //Configure and initialize iflytek services.(This interface must been invoked in application:didFinishLaunchingWithOptions:)
    [IFlySpeechUtility createUtility:initString];
}

#pragma mark - SpeechRecognizer
- (nonnull instancetype)init:(nonnull id<RecognitionListener>)listener {
    if (self = [super init]) {
        self.listener = listener;
        [self setupSpeech:nil];
    }
    return self;
}

- (nonnull instancetype)initWithOptions:(nonnull id<RecognitionListener>)listener options:(nullable NSDictionary<NSString *,id> *)options {
    if (self = [super init]) {
        self.listener = listener;
        [self setupSpeech:options];
    }
    return self;
}

- (void)start:(nullable NSDictionary<NSString *,id> *)options {
    [self setOptions:options];
    [self reset];
    BOOL ret = [_iFlySpeechRecognizer startListening];
    NSLog(@"IFLYSpeechRecognizer: start->%@", ret ? @"YES" : @"NO");
    if (!ret) {
        RecognitionResult *rst = [[RecognitionResult alloc]init];
        rst.errorRst = @"Start speech recognizer failed!";
        [self.listener onError:rst];
    }
}

- (void)cancel {
//    [IFlySpeechRecognizer.sharedcfg cancel];
    [_iFlySpeechRecognizer cancel];
}

- (void)stop {
//    [IFlySpeechRecognizer.sharedcfg stopListening];
    [_iFlySpeechRecognizer stopListening];
}

- (BOOL)destroy {
    if (_iFlySpeechRecognizer) {
        [_iFlySpeechRecognizer cancel];
        return [_iFlySpeechRecognizer destroy];
    }
    return TRUE;
}

#pragma mark - private methods

/**
 * 设置及初始化识别器
 *
 * @param options 需要设置的参数
 */
- (void)setupSpeech:(NSDictionary<NSString*, NSString*> *)options {
    [self createRecognizer];
    [self setOptions:options];
}

/**
 * 创建识别器并设置默认参数
 */
- (void)createRecognizer {
    if (_iFlySpeechRecognizer == nil) {
        _iFlySpeechRecognizer = [IFlySpeechRecognizer sharedInstance];
    }
    [_iFlySpeechRecognizer setParameter:@"" forKey:[IFlySpeechConstant PARAMS]];
    [_iFlySpeechRecognizer setParameter:@"iat" forKey:[IFlySpeechConstant IFLY_DOMAIN]];
    
    IATConfig *cfg = [IATConfig sharedInstance];
    //set timeout of recording
    [_iFlySpeechRecognizer setParameter:cfg.speechTimeout forKey:[IFlySpeechConstant SPEECH_TIMEOUT]];
    //set VAD timeout of end of speech(EOS)
    [_iFlySpeechRecognizer setParameter:cfg.vadEos forKey:[IFlySpeechConstant VAD_EOS]];
    //set VAD timeout of beginning of speech(BOS)
    [_iFlySpeechRecognizer setParameter:cfg.vadBos forKey:[IFlySpeechConstant VAD_BOS]];
    //set network timeout
    [_iFlySpeechRecognizer setParameter:@"20000" forKey:[IFlySpeechConstant NET_TIMEOUT]];
    //set sample rate, 16K as a recommended option
    [_iFlySpeechRecognizer setParameter:cfg.sampleRate forKey:[IFlySpeechConstant SAMPLE_RATE]];
    //set language
    [_iFlySpeechRecognizer setParameter:cfg.language forKey:[IFlySpeechConstant LANGUAGE]];
    //set accent
    [_iFlySpeechRecognizer setParameter:cfg.accent forKey:[IFlySpeechConstant ACCENT]];
    //set whether or not to show punctuation in recognition results
    [_iFlySpeechRecognizer setParameter:cfg.dot forKey:[IFlySpeechConstant ASR_PTT]];
    
    //Set microphone as audio source
    [_iFlySpeechRecognizer setParameter:IFLY_AUDIO_SOURCE_MIC forKey:@"audio_source"];
    //Set result type
    [_iFlySpeechRecognizer setParameter:@"json" forKey:[IFlySpeechConstant RESULT_TYPE]];
//    //Set the audio name of saved recording file while is generated in the local storage path of SDK,by default in library/cache.
//    [_iFlySpeechRecognizer setParameter:@"asr.pcm" forKey:[IFlySpeechConstant ASR_AUDIO_PATH]];
    _iFlySpeechRecognizer.delegate = self;
}

/**
 * 为识别器设置参数
 *
 * @param options 需要设置的参数
 */
- (void)setOptions:(nonnull NSDictionary<NSString*, NSString*> *)options {
    [options enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull key, NSString * _Nonnull obj, BOOL * _Nonnull stop) {
        if ([key isKindOfClass:[NSString class]] && [obj isKindOfClass:[NSString class]]) {
            [_iFlySpeechRecognizer setParameter:obj forKey:key];
        } else {
            // 不支持的类型
            NSLog(@"The key value of an optional parameter must be a numeric value or numeric character");
        }
    }];
}

/**
 * 充值参数
 */
- (void)reset {
    [self cancel];
    self.curResult = [[NSMutableString alloc]init];
}

#pragma mark - IFlySpeechRecognizerDelegate

- (void)onResults:(NSArray *)results isLast:(BOOL)isLast {
    NSMutableString *resultString = [[NSMutableString alloc] init];
    NSDictionary *dic = results[0];
    for (NSString *key in dic) {
        [resultString appendFormat:@"%@",key];
    }
    [self.curResult appendString:[ISRDataHelper stringFromJson:resultString]];
    if (isLast) {
        NSLog(@"IFLYSpeechRecognizer: onResult->%@", self.curResult);
    }
}

- (void)onCompleted:(IFlySpeechError *)error {
    NSLog(@"IFLYSpeechRecognizer: onCompleted->%@", error);
    RecognitionResult *rst = [[RecognitionResult alloc]init];
    rst.type = error.errorCode;
    // 有错误信息
    if (error.errorCode != 0) {
        rst.errorRst = error.errorDesc;
        [self.listener onError:rst];
        return;
    }
    // 无错误信息返回结果
    if (self.curResult.length==0 || [self.curResult hasPrefix:@"nomatch"]) {
        rst.finalRst = @"";
    } else {
        rst.finalRst = _curResult;
    }
    [self.listener onResult:rst];
}

- (void) onVolumeChanged:(int)volume {
//    NSLog(@"IFLYSpeechRecognizer: onVolumeChanged->%d", volume);
//    [self.listener onVolumeChange:volume];
}

- (void) onBeginOfSpeech {
    NSLog(@"IFLYSpeechRecognizer: onBeginOfSpeech");
    [self.listener onRecordingStart];
}

- (void) onEndOfSpeech {
    NSLog(@"IFLYSpeechRecognizer: onEndOfSpeech");
    [self.listener onRecordingStop];
}

- (void) onCancel {
    // do nothing
}

@end
